### PR TITLE
[fix] [broker] response not-found error if topic does not exist when calling getPartitionedTopicMetadata

### PR DESCRIPTION
**Pull request by poorbarcode**
### Motivation & Modifications

Section [Instead of returning a 0 partitioned metadata, respond to a not found error when calling pulsarClient.getPartitionsForTopic(String) if the topic does not exist.](https://github.com/apache/pulsar/blob/master/pip/pip-344.md#goals)

Main PIP: [PIP-344 Correct the behavior of the public API pulsarClient.getPartitionsForTopic(topicName)](https://github.com/apache/pulsar/pull/22182)

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: x

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    private static final Recycler<TopicExistsInfo> RECYCLER = new Recycler<>() {
+        @Override
+        protected TopicExistsInfo newObject(Handle<TopicExistsInfo> handle) {
+            return new TopicExistsInfo(handle);
+        }
+    };
+
+    private static TopicExistsInfo nonPartitionedExists = newInstance(true, TopicType.NON_PARTITIONED, null);
+
+    private static TopicExistsInfo notExists = newInstance(false, TopicType.NON_PARTITIONED, null);
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java:36]: it would be better to instantiate these instances directly with a special constructor so that `Recycler.Handle` is set to null. This would prevent accidential call to `.recycle`. There's no need to have the validation of the `newInstance` method for these cases.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java [hunk 2]
```diff
@@ -752,6 +752,14 @@ public CompletableFuture<Void> closeAsync() {
...
         List<CompletableFuture<Void>> futures = new ArrayList<>();
 
+        urlLookupMap.entrySet().forEach(e -> {
+            try {
+                e.getValue().close();
+            } catch (Exception ex) {
+                log.error("Error closing lookup service {}", e.getKey(), ex);
+            }
+        });
+
```
Comment (lhotari) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java:762]: It would be worth creating a separate PR for this fix. The benefit is that it could be cherry-picked to older maintenance branches which won't be getting this current PR.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java (hunk 1 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java]: Improved.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java (hunk 2 above)
Comment (poorbarcode) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java]: Created a separate PR for this fix
- https://github.com/apache/pulsar/pull/22858

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 3]
```diff
@@ -1400,40 +1403,100 @@ public CompletableFuture<List<String>> getOwnedTopicListForNamespaceBundle(Names
...
+            }
+            LookupOptions lookupOptions = LookupOptions.builder().readOnly(false).authoritative(true).build();
+            return getBrokerServiceUrlAsync(TopicName.get(topic), lookupOptions)
+                .thenCompose(lookupResult -> {
+                    if (!lookupResult.isPresent()) {
+                        log.error("{} Failed to get partition metadata due can not find the owner broker", topic);
+                        return FutureUtil.failedFuture(new ServiceUnitNotReadyException(
+                                "No broker was available to own " + topicName));
+                    }
+                    return pulsarClient.getLookup(lookupResult.get().getLookupData().getBrokerUrl())
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:1485]: Can we just return 307 to the client side for forwarding? 
Now, the solution is a proxy. 

Actually, we already have the topic ownership checking in non-persistent topic lookup. Keeping them consistent will be better for code maintenance.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            return new TopicExistsInfo(handle);
+        }
+    };
+
+    private static TopicExistsInfo nonPartitionedExists = new TopicExistsInfo(null);
+    static {
+        nonPartitionedExists.exists = true;
+        nonPartitionedExists.topicType = TopicType.NON_PARTITIONED;
+        nonPartitionedExists.partitions = null;
+    }
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java:39]: Is it better just set the default values with the constructor? Just keep them consistent as how do you set the `handler`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+    private static TopicExistsInfo notExists = new TopicExistsInfo(null);
+    static {
+        notExists.exists = false;
+        notExists.topicType = TopicType.NON_PARTITIONED;
+        notExists.partitions = null;
+    }
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java:46]: Same as the above comment.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    public static TopicExistsInfo partitionedExists(Integer partitions){
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java:48]:
```suggestion
    public static TopicExistsInfo newPartitionedTopicExists(Integer partitions){
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        return newInstance(true, TopicType.PARTITIONED, partitions);
+    }
+
+    public static TopicExistsInfo nonPartitionedExists(){
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java:52]:
```suggestion
    public static TopicExistsInfo newNonPartitionedTopicExists(){
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        return nonPartitionedExists;
+    }
+
+    public static TopicExistsInfo notExists(){
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java:56]:
```suggestion
    public static TopicExistsInfo newTopicNotExists(){
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            }
+            info.partitions = partitions.intValue();
+        } else {
+            if (partitions != null) {
+                throw new IllegalArgumentException("The param partitions must be null when creating a non-partitioned"
+                        + " TopicExistsInfo obj.");
+            }
+        }
+        return info;
+    }
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java:81]: Actually, we only need to create instance for partitioned topic.
I think we can just add the implementation in the `newPartitionedTopicExists` method.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+    private final Recycler.Handle<TopicExistsInfo> handle;
+
+    @Getter
+    private TopicType topicType;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java:86]: It looks like we don't really need the topicType field.
We can just keep partitions as zero by default and create a method `getTopicType()` and return PartitionedTopic if partitions > 0;

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    @Getter
+    private Integer partitions;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java:88]: `int` should be enough for this case? We don't really need a null value.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    @Getter
+    private boolean exists;
+
+    private TopicExistsInfo(Recycler.Handle<TopicExistsInfo> handle) {
+        this.handle = handle;
+    }
+
+    public void recycle(){
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java:96]:
```suggestion
    public void recycle() {
```

Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java]: Fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java (hunk 11 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java]: Change the type to `int`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java (hunk 10 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java]: Removed the field `topicType`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java (hunk 9 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java]: Improved

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java (hunk 8 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java]: Renamed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java (hunk 7 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java]: Renamed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java (hunk 6 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java]: Renamed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java (hunk 5 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java]: Improved

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java (hunk 4 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/TopicExistsInfo.java]: Improved

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java [hunk 13]
```diff
@@ -70,13 +70,14 @@ protected CompletableFuture<LookupData> internalLookupTopicAsync(final TopicName
...
-
-                    return existFuture;
+                    return pulsar().getNamespaceService().checkTopicExists(topicName, true).thenCompose(info -> {
+                        boolean exists = info.isExists();
+                        info.recycle();
+                        if (exists) {
+                            return CompletableFuture.completedFuture(true);
+                        }
+                        return pulsar().getBrokerService().isAllowAutoTopicCreationAsync(topicName);
+                    });
```
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java:80]: > Instead, we can just return true here if the topic is a non-persistent topic.

After this change, the below three behaviors 

1. Disable auto-creation, the `non-persistent` topic does not exist
  a. response: `{partitions: 0}`, but the expected behavior should be `Topic No Found`. In my opinion, the response `{partitions: 0}` is also okay, because it is the original behavior before the current PR.
2. Enable auto-creation; `allowAutoTopicCreationType=partitioned`; `defaultNumPartitions: 3`. 
  a. response: `{partitions: 0}`, but the expected behavior should be `{partitions: 3}`. In my opinion, `{partitions: 0}` is better than `{partitions: 3}`, because it can avoid auto create partitioned non-persistent topics.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java (hunk 3 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:1487]: > Instead, we can just return true here if the topic is a non-persistent topic.

After this change, the method `checkNonPersistentNonPartitionedTopicExists(String topic)` will be simpler as below, so we can discuss to [the above comment](https://github.com/apache/pulsar/pull/22838#discussion_r1634259488) first.

```java
private CompletableFuture<Boolean> checkNonPersistentNonPartitionedTopicExists(String topic) {
  return CompletableFuture.completedFuture(true);
}

```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java]: Sorry, I misunderstood before, the param `assumedNonPartitionedNonPersistentTopicAlwaysExists` of `NamespaceService.checkTopicExists()` has been removed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java (hunk 3 above)
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:1471]: Sorry, I misunderstood before.

> Can we just return 307 to the client side for forwarding?

So far, the client with binary protocol does not support redirect response code, so we do the redirect in the broker internally is better.

Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:1471]: Ok, I see.

**Review by codelipenghui — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java [hunk 14]
```diff
@@ -104,6 +104,29 @@ public Consumer<T> subscribe() throws PulsarClientException {
         }
     }
 
+    private CompletableFuture<Boolean> topicExists(String topic) {
+        CompletableFuture<Boolean> existsFuture = new CompletableFuture<>();
+        client.getPartitionedTopicMetadata(topic, false).thenAccept(metadata -> {
+            TopicName topicName = TopicName.get(topic);
+            if (topicName.isPersistent()) {
+                existsFuture.complete(true);
```
Comment (Technoboy-) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java:112]: If the topic is persistent, so don't need to get the metadata ?

Comment (poorbarcode) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java:113]: This method is used to check the topic created for the old `DLQ` rule(`{namespace}-{subscription}-DLQ`) already exists
- if yes, use the old topic
- else, create DLQ with the new rule(`{topic}-{subscription}-DLQ`)

So if it is a persistent topic, the result is `true` whether it is partitioned or not. If it is a non-persistent topic, the result is `false` if it is a non-partitioned topic(the non-partitioned DLQ will be removed automatically after the consumers are offline)

Comment (poorbarcode) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java:113]: I renamed the method to make it more clear

Comment (Technoboy-) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java:113]: ok

**Review by Technoboy- — APPROVED** [review]
